transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java

```diff
@@ -0,0 +1,200 @@
+ * under the License.
+ */
+package io.netty.channel.pool;
+
+import io.netty.bootstrap.Bootstrap;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.EventLoop;
+import io.netty.util.AttributeKey;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.GenericFutureListener;
+import io.netty.util.concurrent.ImmediateEventExecutor;
+import io.netty.util.concurrent.Promise;
+import io.netty.util.internal.PlatformDependent;
+
+import java.util.Deque;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.LinkedBlockingDeque;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+
+/**
+ * Simple {@link ChannelPool} implementation which will create new {@link Channel}s if someone tries to acquire
+ * a {@link Channel} but none is in the pool atm. No limit on the maximal concurrent {@link Channel}s is enforced.
+ *
+ * @param <C>   the {@link Channel} type to pool.
+ * @param <K>   the {@link ChannelPoolKey} that is used to store and lookup the {@link Channel}s.
+ */
+public final class SimpleChannelPool<C extends Channel, K extends ChannelPoolKey> implements ChannelPool<C, K> {
+
+    private static final AttributeKey<ChannelPoolKey> KEY = AttributeKey.newInstance("channelPoolKey");
+    private final ConcurrentMap<ChannelPoolKey, Deque<C>> pool = PlatformDependent.newConcurrentHashMap();
+    private final ChannelPoolHandler<C, K> handler;
+    private final ChannelPoolHealthChecker<C, K> healthCheck;
+    private final Bootstrap bootstrap;
+
+    public SimpleChannelPool(Bootstrap bootstrap, final ChannelPoolHandler<C, K> handler) {
+        this(bootstrap, handler, ActiveChannelPoolHealthChecker.<C, K>instance());
+    }
+
+    public SimpleChannelPool(Bootstrap bootstrap, final ChannelPoolHandler<C, K> handler,
+                             final ChannelPoolHealthChecker<C, K> healthCheck) {
+        this.handler = checkNotNull(handler, "handler");
+        this.healthCheck = checkNotNull(healthCheck, "healthCheck");
+        this.bootstrap = checkNotNull(bootstrap, "bootstrap").clone();
+        this.bootstrap.handler(new ChannelInitializer<C>() {
+            @SuppressWarnings("unchecked")
+            @Override
+            protected void initChannel(C ch) throws Exception {
+                K key = (K) ch.attr(KEY).get();
+                assert key != null;
+                handler.channelCreated(ch, key);
+            }
+        });
+    }
+
+    @Override
+    public Future<C> acquire(K key) {
+        return acquire(key, ImmediateEventExecutor.INSTANCE.<C>newPromise());
+    }
+
+    @Override
+    public Future<C> acquire(final K key, final Promise<C> promise) {
+        checkNotNull(key, "key");
+        checkNotNull(promise, "promise");
+        try {
+            Deque<C> channels = pool.get(key);
```

Comment:
This doesn't sound right to me.  Let's say you acquires and release a channel with a key (addr1, loopA). Later, when you try to acquire a channel with a key (addr1, null), you cannot reuse the channel.